File: 0081-Search-in-Rotated-Sorted-Array-II/lc0081.cc
// C++ solution.

// Very similar to LC33, just requires an additional duplicate removal.
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        while (low < high && nums[low] == nums[high]) {
            low++;
        }
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target) {
                return true;
            }
            if (nums[mid] > target) {
                if (nums[mid] < nums[low] || target >= nums[low]) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else {
                if (nums[mid] >= nums[low] || target < nums[low]) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
        }
        return false;
    }
};

File: 0081-Search-in-Rotated-Sorted-Array-II/lc0081.rs
// Rust solution.

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> bool {
        if !nums.is_empty() {
            let mut head: usize = 0;
            let n: usize = nums.len();
            while head < n-1 && nums[head] == nums[n-1] {
                head += 1;
            }

            let (mut low, mut high) = (head, n);
            while low < high {
                let mid = (low + high) >> 1;
                if nums[mid] == target {
                    return true;
                } else if !((nums[head] <= target) ^ (nums[head] > nums[mid]) ^ (nums[mid] > target)) {
                    high = mid;
                } else {
                    low = mid+1;
                }
            }
        }
        false
    }
}
